tests/feature.test.js

```diff
@@ -1,79 +1,286 @@
+  console.log = function () {}
+  console.warn = function () {}
+
+  beforeEach(() => {
+    jest.spyOn(console, 'error')
+    jest.spyOn(console, 'log')
+    jest.spyOn(console, 'warn')
+  })
+  afterEach(() => {
+    jest.resetModules()
+  })
+  afterAll(() => {
+    jest.clearAllMocks()
   })
 
-  it('expects isEqual to correctly equate two multi value features', () => {
-    let a = new Feature(Feature.types.note, ['valuea', 'valueb'], Constants.LANG_LATIN)
-    let b = new Feature(Feature.types.note, ['valuea', 'valueb'], Constants.LANG_LATIN)
-    expect(a.isEqual(b)).toBeTruthy()
+  let latID = Constants.LANG_LATIN
+
+  it('1 Feature - check required arguments for Feature constructor', () => {
+    expect(function () {
+      let l = new Feature('fooType')
+      console.log(l)
+    }).toThrow(new Error('Features of "fooType" type are not supported.'))
+
+    expect(function () {
+      let l = new Feature('word')
+      console.log(l)
+    }).toThrow(new Error('Feature should have a non-empty value(s).'))
+
+    expect(function () {
+      let l = new Feature('word', {})
+      console.log(l)
+    }).toThrow(new Error('No language ID is provided'))
   })
 
-  it('expects isEqual to correctly not equate two single value features', () => {
-    let a = new Feature(Feature.types.note, 'valuea', Constants.LANG_LATIN)
-    let b = new Feature(Feature.types.note, 'valueb', Constants.LANG_LATIN)
-    expect(a.isEqual(b)).toBeFalsy()
+  it('2 Feature - create with min arguments', () => {
+    let feature = new Feature('word', 'fooword', latID)
+
+    expect(feature.type).toEqual('word')
+    expect(feature.languageID).toEqual(latID)
+    expect(feature.sortOrder).toEqual(1)
+    expect(feature.allowedValues).toEqual([])
+
+    expect(feature._data).toEqual([ { value: 'fooword', sortOrder: 1 } ])
+
+    expect(feature.allowsUnrestrictedValues).toBeTruthy()
+    expect(feature.valuesUnrestricted).toBeTruthy()
   })
 
-  it('expects isEqual to correctly not equate two multi value features', () => {
-    let a = new Feature(Feature.types.note, ['valuea', 'valueb'], Constants.LANG_LATIN)
-    let b = new Feature(Feature.types.note, ['valuea', 'valuec'], Constants.LANG_LATIN)
-    expect(a.isEqual(b)).toBeFalsy()
+  it('3 Feature - create with full arguments', () => {
+    let feature = new Feature('word', 'fooword', latID, 10, ['val1', 'val2'])
+
+    expect(feature.sortOrder).toEqual(10)
+    expect(feature.allowedValues).toEqual(['val1', 'val2'])
+
+    expect(feature.allowsUnrestrictedValues).toBeFalsy()
+    expect(feature.valuesUnrestricted).toBeFalsy()
   })
 
-  it('expects isEqual to correctly not equate two single value features of different type', () => {
-    let a = new Feature(Feature.types.note, 'valuea', Constants.LANG_LATIN)
-    let b = new Feature(Feature.types.frequency, 'valuea', Constants.LANG_LATIN)
-    expect(a.isEqual(b)).toBeFalsy()
+  it('4 Feature - check dataValuesFromInput, sort, compareTo, items, isEqual', () => {
```

Comment:
I'm not sure if all of the equality and sorting tests that were included in the prior version are covered now. Is there a reason to remove them?

Reply:
I have changed the way of creating and going through the tests to increase coverage and it is more simple to create them from scratch.

And I believe this approach allows to track dependent changes easier.
For example - for comparing I have used isEqual and CompareTo in one unit case on the same features. And it goes through all code branches and checks.

I could add your test cases to mine - if you doubt of the resulted quality - no problems with it :)

Reply:
Thank you :-) I'm not positive they weren't covered, just didn't immediately see how they were.

Reply:
I have looked more attentively on your tests and they are testing some cases that cover the same code from one point of view. But as for the future development and for future code changes - they will be really important.
So you were right - my tests were not as detailed as yours and they didn't cover all these variants listed in your tests.
So I have added them with a pleasure.
Thank you for being so attentive!